presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftHiveMetastore.java

```diff
@@ -771,7 +788,37 @@ public void alterTable(String databaseName, String tableName, Table table)
     }
 
     @Override
-    public void addPartitions(String databaseName, String tableName, List<Partition> partitions)
+    public void addPartitions(String databaseName, String tableName, List<PartitionWithStatistics> partitionsWithStatistics)
+    {
+        List<Partition> partitions = partitionsWithStatistics.stream()
+                .map(ThriftMetastoreUtil::toMetastoreApiPartition)
+                .collect(toImmutableList());
+        addMetastorePartitions(databaseName, tableName, partitions);
+        try {
+            if (partitionsWithStatistics.stream().allMatch(p -> p.getStatistics().getColumnStatistics().isEmpty())) {
+                return;
+            }
+            for (PartitionWithStatistics partitionWithStatistics : partitionsWithStatistics) {
+                storePartitionColumnStatistics(databaseName, tableName, partitionWithStatistics.getPartitionName(), partitionWithStatistics);
+            }
+        }
+        catch (Throwable t) {
+            for (Partition partition : partitions) {
+                try {
+                    dropPartition(databaseName, tableName, partition.getValues(), false);
+                }
+                catch (Throwable suppressed) {
+                    if (t != suppressed) {
+                        t.addSuppressed(suppressed);
+                    }
+                }
+            }
+            throw t;
+        }
+    }
+
+    @Override
+    public void addMetastorePartitions(String databaseName, String tableName, List<Partition> partitions)
```

Comment:
I'm going to remove the legacy path once we see that the new path works fine. Hopefully it is going to be removed in a very next release. I wouldn't be bother much about it. I can rename it though.

Reply:
I think this is reasonable. Add a TODO. It will serve two purposes:

* remind us to remove the code
* (implicitly and effectively) tell other people to not remove the code despite it's apparently unused.